### PR TITLE
augmented utxo view for steady state & syncing stack

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
Current dependencies on/for this PR:
* `feature/pos-syncing-and-steady-state`
  * **PR #987** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/987?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/987?utm_source=stack-comment).

File: lib/pos_mempool.go [hunk 1]
```diff
@@ -129,6 +140,14 @@ type PosMempool struct {
 	// feeEstimator is used to estimate the fee required for a transaction to be included in the next block
 	// based off the current state of the mempool and the most n recent blocks.
 	feeEstimator *PoSFeeEstimator
+
+	// augmentedBlockViewRefreshIntervalMillis is the frequency with which the augmentedLatestBlockView is updated.
+	augmentedBlockViewRefreshIntervalMillis int64
```
Comment (tholonious) [inline lib/pos_mempool.go:145]:
```suggestion
	augmentedBlockViewRefreshIntervalMillis uint64
```

File: lib/pos_mempool.go [hunk 2]
```diff
@@ -129,6 +140,14 @@ type PosMempool struct {
...
+
+	// augmentedLatestBlockViewSequenceNumber is the sequence number of the readOnlyLatestBlockView. It is incremented
+	// every time augmentedLatestBlockView is updated. It can be used by obtainers of the augmentedLatestBlockView to
+	// wait until a particular transaction has been connected.
+	augmentedLatestBlockViewSequenceNumber int64
```
Comment (tholonious) [inline lib/pos_mempool.go:150]:
```suggestion
	augmentedLatestBlockViewSequenceNumber uint64
```

File: lib/pos_mempool.go [hunk 3]
```diff
@@ -179,6 +199,7 @@ func (mp *PosMempool) Init(
 	feeEstimatorNumMempoolBlocks uint64,
 	feeEstimatorPastBlocks []*MsgDeSoBlock,
 	feeEstimatorNumPastBlocks uint64,
+	augmentedBlockViewRefreshIntervalMillis int64,
```
Comment (tholonious) [inline lib/pos_mempool.go:202]:
```suggestion
	augmentedBlockViewRefreshIntervalMillis uint64,
```

File: lib/pos_mempool.go [hunk 4]
```diff
@@ -241,11 +270,72 @@ func (mp *PosMempool) Start() error {
...
 		}
 	}
+	mp.startAugmentedViewRefreshRoutine()
 
 	mp.status = PosMempoolStatusRunning
 	return nil
 }
 
+func (mp *PosMempool) startAugmentedViewRefreshRoutine() {
+
```
Comment (tholonious) [inline lib/pos_mempool.go:280]:
```suggestion
```

File: lib/pos_mempool.go [hunk 5]
```diff
@@ -241,11 +270,72 @@ func (mp *PosMempool) Start() error {
...
+					}
+				}
+				// Grab the augmentedLatestBlockViewMutex write lock and update the augmentedLatestBlockView.
+				mp.augmentedLatestBlockViewMutex.Lock()
+				mp.augmentedLatestBlockView = newView
+				mp.augmentedLatestBlockViewMutex.Unlock()
+				// Increment the augmentedLatestBlockViewSequenceNumber.
+				atomic.AddInt64(&mp.augmentedLatestBlockViewSequenceNumber, 1)
+			case <-mp.quit:
+				break out
```
Comment (tholonious) [inline lib/pos_mempool.go:333]: nit: you can just just return, no?
```suggestion
				return
```

File: lib/pos_mempool.go [hunk 6]
```diff
@@ -241,11 +270,72 @@ func (mp *PosMempool) Start() error {
...
+	mp.startAugmentedViewRefreshRoutine()
 
 	mp.status = PosMempoolStatusRunning
 	return nil
 }
 
+func (mp *PosMempool) startAugmentedViewRefreshRoutine() {
+
+	go func() {
+	out:
```
Comment (tholonious) [inline lib/pos_mempool.go:282]:
```suggestion
```

**Review by tholonious — APPROVED** [review]

**Comment by lazynina** [comment]
### Merge activity

* **Jan 31, 5:43 PM EST America/New_York**: @lazynina started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/987).

**Pull request merged**